AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m

```diff
@@ -188,6 +188,137 @@ - (void)enqueueItem:(id<MSLog>)item flags:(MSFlags)flags {
+                                }
+                              }];
+
+    // Forward logs to the ingestion.
+    [self.ingestion
+        sendAsync:container
+        authToken:tokenInfo.authToken
+completionHandler:^(NSString *ingestionBatchId, NSHTTPURLResponse *response, __attribute__((unused)) NSData *data, NSError *error) {
+  dispatch_async(self.logsDispatchQueue, ^{
+    if ([self.pendingBatchIds containsObject:ingestionBatchId]) {
+
+      // Success.
+      if (response.statusCode == MSHTTPCodesNo200OK) {
+        MSLogDebug([MSAppCenter logTag], @"Log(s) sent with success, batch Id:%@.", ingestionBatchId);
+
+        // Notify delegates.
+        [self enumerateDelegatesForSelector:@selector(channel:didSucceedSendingLog:)
+                                  withBlock:^(id <MSChannelDelegate> delegate) {
+                                    for (id <MSLog> aLog in logArray) {
+                                      [delegate channel:self didSucceedSendingLog:aLog];
+                                    }
+                                  }];
+
+        // Remove from pending logs and storage.
+        [self.pendingBatchIds removeObject:ingestionBatchId];
+        [self.storage deleteLogsWithBatchId:ingestionBatchId groupId:self.configuration.groupId];
+
+        // Try to flush again if batch queue is not full anymore.
+        if (self.pendingBatchQueueFull && self.pendingBatchIds.count < self.configuration.pendingBatchesLimit) {
+          self.pendingBatchQueueFull = NO;
+          if (self.availableBatchFromStorage) {
+            [self sendLogArray:logArray
+                   withBatchId:batchId
+                  andAuthToken:tokenInfo];
+          }
+        }
+      } else {
+        MSLogError([MSAppCenter logTag], @"Log(s) sent with failure, batch Id:%@, status code:%tu", ingestionBatchId,
+          response.statusCode);
+
+        // Notify delegates.
+        [self enumerateDelegatesForSelector:@selector(channel:didFailSendingLog:withError:)
+                                  withBlock:^(id <MSChannelDelegate> delegate) {
+                                    for (id <MSLog> aLog in logArray) {
+                                      [delegate channel:self didFailSendingLog:aLog withError:error];
+                                    }
+                                  }];
+
+        // Remove from pending logs.
+        [self.pendingBatchIds removeObject:ingestionBatchId];
+        [self.storage deleteLogsWithBatchId:ingestionBatchId groupId:self.configuration.groupId];
+
+        // Update pending batch queue state.
+        if (self.pendingBatchQueueFull && self.pendingBatchIds.count < self.configuration.pendingBatchesLimit) {
+          self.pendingBatchQueueFull = NO;
+        }
+      }
+    } else {
+      MSLogWarning([MSAppCenter logTag], @"Batch Id %@ not expected, ignore.", ingestionBatchId);
+    }
+  });
+}];
+  }
+}
+
+- (void)flushQueueBatchForTokenArray:(NSMutableArray<MSAuthTokenHistoryState *> *)tokenArray
```

Comment:
As for the name - yeah, that's a good point; As for the loop - unfortunatelly, no, even though that would reduce the amount of time needed to read the code. Currently we have 2 calls to recursion and async task in a callback. Plain iteration won't do the trick here: we don't know the result of sending on-demand (think of it as in the loop iteration code) and can only get the result in async task's callback. Thus, we can only use another recursive call rather than loop